### ISSUE TITLE
Atomic species flag for deuterium (H2) is replaced by oxygen (O) on Quantum ESPRESSO inputs

### ISSUE THREAD

**Issue by pablogila** [phonopy/phonopy#412]
Hello there,  

I am running some Phonopy calculations with a deuterated material. Quantum ESPRESSO does not understand `D` at atomic species, but instead we need an `H2` flag such as follows:

```
ATOMIC_SPECIES
   I  126.90400   I.upf
   H    1.00750   H.upf
   N   14.00650   N.upf
  H2    2.01410   H.upf
  Pb  207.20000  Pb.upf
   C   12.01060   C.upf
```

The problem comes when I run Phonopy to create the supercells, as follows:

```shell
phonopy --qe -d --dim="2 2 2" -c scf.in --tolerance=1e-3
```

then Phonopy replaces this `H2` flag by an `O` in all the supercells for the scf calculation, without previous warning:

```
ATOMIC_SPECIES
 Pb  207.20000   Pb.upf
  I  126.90400   I.upf
  N   14.00650   N.upf
  C   12.01060   C.upf
  O    2.01410   H.upf
  H    1.00750   H.upf
```

This causes problems down the line to extract theoretical Inelastic Neutron Scattering spectra from the Phonopy output, since for this matter oxygen plays quite different than my beloved deuterium.

**Comment by atztogo** [comment]
1. Could you give `scf.in` to reproduce this issue?
2. What is the definition of the structure format containing `H2`?
3. What is your expected behaviour? Please explain more in details.

**Comment by pablogila** [comment]
Down below is the `scf.in` file.

`H2` is the atomic species for deuterium in Quantum ESPRESSO, since it has to be a former chemical symbol plus a number (from [ATOMIC_SPECIES documentation](https://www.quantum-espresso.org/Doc/INPUT_PW.html#ATOMIC_SPECIES)). QE does not understand non-standard symbols like `D`.

After creating the supercells as:
```
phonopy --qe -d --dim="2 2 2" -c scf.in --tolerance=1e-3
```
the expected behaviour would be that these `H2` flags for deuterium would remain in the supercell inputs. However, this is not the case, and all `H2` references are replaced by oxygens `O`, both in the `ATOMIC_SPECIES` and `ATOMIC_POSITIONS` cards.

```
! scf.in file

&CONTROL
  calculation = 'scf'
  prefix = 'MAPbI3'
  outdir = './tmp/'
  pseudo_dir = './pseudodojo-pbesol-paw-standard/'
  restart_mode = 'from_scratch'
  max_seconds = 85000
  etot_conv_thr = 3.5279349288639156d-12
  forc_conv_thr = 3.889380759660317d-07
  tprnfor = .true.
  tstress = .true.
  nstep = 9999
/

&SYSTEM
  ecutwfc = 66.14877991619842
!  ecutrho = ! 4 * ecutwfc
  ibrav = 0
!  A =   12.93110
  celldm(1) = 24.43623749
!  nat = 384
  nat = 48
  ntyp = 6
  input_dft = 'PBEsol'
  vdw_corr = 'DFT-D3'
  dftd3_version = 4
  dftd3_threebody = .true.
  occupations = 'fixed'
/

&ELECTRONS
  adaptive_thr = .true.
  conv_thr = 3.5279349288639156d-12
  diagonalization = 'ppcg'
!  mixing_beta = 0.7d0
  diago_thr_init = 1.0d-7
  electron_maxstep = 9999
/

&IONS
  ion_dynamics = 'bfgs'
  trust_radius_min = 1.8897261259077822d-06
  upscale = 10000
/

&CELL
  cell_dynamics = 'bfgs'
!  cell_dofree = 'volume'
  press = 0.0001
  press_conv_thr = 0.0001
/

K_POINTS (automatic)
  2 2 2 1 1 1

!BEGIN_COORDINATES

ATOMIC_SPECIES
   I  126.90400   I.upf
   H    1.00750   H.upf
   N   14.00650   N.upf
  H2    2.01410   H.upf
  Pb  207.20000  Pb.upf
   C   12.01060   C.upf

CELL_PARAMETERS alat
   0.490462062  -0.466047333   0.000000000
   0.490462062   0.466047333   0.000000000
   0.000000000   0.000000000   0.935919893

ATOMIC_POSITIONS (crystal)
Pb            0.0000000000        0.0000000000       -0.0000000000
Pb           -0.0000000000       -0.0000000000        0.5000000000
Pb            0.5000000000        0.5000000000       -0.0000000000
Pb            0.5000000000        0.5000000000        0.5000000000
I             0.2504501583        0.2504501583       -0.0000000000
I             0.2504501583        0.2504501583        0.5000000000
I             0.7495498417        0.7495498417        0.5000000000
I             0.7495498417        0.7495498417        0.0000000000
I             0.0545341167        0.9454658833        0.7500000000
I             0.9454658833        0.0545341167        0.2500000000
I             0.4316207445        0.5683792555        0.7500000000
I             0.5683792555        0.4316207445        0.2500000000
I             0.7503142126        0.2496857874        0.9258551107
I             0.7503142126        0.2496857874        0.5741448893
I             0.2496857874        0.7503142126        0.4258551107
I             0.2496857874        0.7503142126        0.0741448893
N             0.3677255518        0.0526094646        0.2500000000
N             0.0526094646        0.3677255518        0.7500000000
N             0.6322744482        0.9473905354        0.7500000000
N             0.9473905354        0.6322744482        0.2500000000
C             0.5289907860        0.9943647985        0.2500000000
C             0.9943647985        0.5289907860        0.7500000000
C             0.4710092140        0.0056352015        0.7500000000
C             0.0056352015        0.4710092140        0.2500000000
H2            0.6585700836        0.8807333227        0.8197201612
H2            0.6585700836        0.8807333227        0.6802798388
H2            0.8807333227        0.6585700836        0.3197201612
H2            0.8807333227        0.6585700836        0.1802798388
H2            0.3414299164        0.1192666773        0.3197201612
H2            0.1192666773        0.3414299164        0.6802798388
H2            0.3414299164        0.1192666773        0.1802798388
H2            0.1192666773        0.3414299164        0.8197201612
H             0.5501242558        0.9243176149        0.3243519922
H             0.5501242558        0.9243176149        0.1756480078
H             0.9243176149        0.5501242558        0.8243519922
H             0.9243176149        0.5501242558        0.6756480078
H             0.4498757442        0.0756823851        0.8243519922
H             0.0756823851        0.4498757442        0.1756480078
H             0.4498757442        0.0756823851        0.6756480078
H             0.0756823851        0.4498757442        0.3243519922
H             0.6029210884        0.0920390449        0.2500000000
H             0.0920390449        0.6029210884        0.7500000000
H             0.3970789116        0.9079609551        0.7500000000
H             0.9079609551        0.3970789116        0.2500000000
H2            0.0374027550        0.7057687624        0.2500000000
H2            0.7057687624        0.0374027550        0.7500000000
H2            0.9625972450        0.2942312376        0.7500000000
H2            0.2942312376        0.9625972450        0.2500000000
```

**Comment by atztogo** [comment]
Are `H`, `H1`, `H2`, `H3` treated as different species? Did you choose `H2` although it could be `H1`?

This logic is not considered and not implemented in phonopy, therefore it behaved strangely.

So you think phonopy should support this feature and you ask me to implement it. Am I correct?

**Comment by pablogila** [comment]
Quantum ESPRESSO treats atoms according to their mass and pseudopotential. Adding numbers such as `H2` is supported to differentiate atoms, e.g. different isotopes, but that's it. As long as the mass and pseudo remain the same, the results would be the same. If I had just called it `H3`, QE would give the same result. Without this ability to add numbers after the atomic species, there would be no way to differentiate between protium and deuterium in the same calculation, which is my case.

One important point: Once the QE' SCF calculation is finished and the forces are calculated with
```
phonopy -f supercell-*.out
phonopy --include-all
```
The resulting `phonopy.yaml` should have `H`, not `H2`, since the atomic mass is already specified and the custom atomic species is no longer needed.

I have tried to customise `/phonopy/interface/qe.py` and `/phonopy/structure/atoms.py`, but I lack the knowledge of this code to avoid creating more bugs than I want to solve. So if you could implement this feature, it would be more than welcome.
Thanks :D

**Comment by atztogo** [comment]
Thanks for your explanation. My question was about how the crystal symmetry is defined when using `H` and `H1` simultaneously. For example, we have a diamond structure with 8 `C` atoms (conventional unit cell). If we replace one atom by `C1`, do you get the same set of space group operations or not?

**Comment by pablogila** [comment]
In this specific case we have the same space group operations. This material contains methylammonium (CH3NH3). We are selectively deuterating only the C or the N (as in CD3NH3 or CH3ND3), and here the symmetry was already predefined by positions of the C and N.

**Comment by atztogo** [comment]
Your explanation is not clear enough to me. Do you mean, QE tries to search space group operations by treating `C` and `C1`, `C2`, ... equally? Or `C` and `C2` are considered as different species, and the symmetry can be broken with respect to the crystal structure with all `C`?

Two issues:
1. Phonopy doesn't support reading `H2` like symbol. To support it, it will need a design choice around symmetry treatment. Like as you wrote, if the set of space group operations will not be changed by using `H2` instead of `H`, you can simply use mass tag (https://phonopy.github.io/phonopy/setting-tags.html#mass)
2. Since `H2` is not supported, phonopy behaved strangely like the symbol became `O`. This can be captured as an exception.

Currently I am not sure what I should do.

**Comment by pablogila** [comment]
Sorry, I did not explain properly. QE considers `H` and `H2` (or `C` and `C3`, etc) to be **different** species, and so the symmetry can indeed be changed. Which is not my particular case, but it could happen anyway with a different material.  

I observed several things that may help in choosing a workaround:  

- The mass of the `H2` indicated in the `scf.in` was preserved after creating the supercells, even though it was marked as `O`. It was not replaced by the oxygen mass, nor the protium one.  
- Phonopy detected the symmetry properly, considering the newly-replaced oxygens as different atomic species.  
- With other materials it was not `O` but `F`, so the new rogue atomic species may vary.  
- This replacement of the atomic species also takes place when generating the forces with `phonopy -f supercell-*.out` and `phonopy --include-all`.  

In principle, these SCF calculations to get the phonons are only affected by the assigned pseudopotentials and masses. I did 2 test runs: one by calculating the phonons with the rogue atomic species, and a second one by replacing all the rogue `O` atoms in the generated supercells by the desired `H2` atoms. In both cases I had to modify the final `phonopy.yaml` to indicate the desired `H2` atomic species. The results were apparently the same, and the phonons were calculated correctly in both cases.  

Since the mass was preserved, I assume that the new standard masses are assigned by Phonopy **before** the atomic species replacement, and since Phonopy has no `H2` mass assigned, it just uses the mass from the input. This detail allows us to run the calculations properly despite the atomic replacement.  

I can think of a possible workaround, of course without any further knowledge of the code internals:  

Whenever an unknown species is detected, it can continue to do what it is doing. When it has finished, it could check whether the atomic species in the new supercells have been changed. If so, it could re-read the input and assign the old, unknown atomic species from the input to the new, rogue ones that were not there before. If there were more than one of these unknown atomic species, it could assign the correct values by checking that their masses are the same.  

I have coded some bash scripts to perform these corrections manually, available at https://github.com/pablogila/phonomake (see `pm_fix_yaml.sh`). I have not included the correction of the generated supercells, as the results seem to be the same with and without the rogue species, although I may upload it later, just in case. I am considering automating this workaround with Python, although I don't know if it would make sense to integrate this approach into Phonopy nor how to do it properly. What do you think?

**Comment by atztogo** [comment]
I wrote the qe interface 10 years ago. I read it now (the code is not nicely written...), I found two things:

1. As you wrote, masses are read from `scf.in` when phonopy finds a chemical symbol that is not in the list defined in phonopy. In your case, `H2` applies to that. None of those are found, masses that are implemented in phonopy are used.
2. When phonopy finds the unknown chemical symbol like `H2`, then that is replaced a chemical symbol that is not used in `scf.in`, but the mass in `scf.in` is used.

So the issue is probably about how to display chemical symbols. To handle this `PhonopyAtoms` has to be modified to allow user defined chemical symbols and their masses.

Supporting all requirement of QE
```
"Xn" (e.g. Fe1) or "X_*" or "X-*"
```
requires a big job. If user's custom definitions are allowed, those information has to be copied when generating new `PhonopyAtoms` instance based on other `PhonopyAtoms` instance. This means all the lines doing this have to be updated, and it can induce bugs. So we need compromise. Do you think QE users can be happy only supporting `"Xn"` with `n>0`? In this case, I feel that the implementation may be achievable.

Phonopy is used by a huge number of people, but the number of the developers who maintain the code is few. Implementation of additional user interface increases maintenance costs. If the maintenance costs are not estimated correctly, it will become unsustainable, which is not good for users. So we developers always have to consider about the resource required for the maintenance. This is one of the reasons why we can't fulfill all the wishes.

**Comment by atztogo** [comment]
I pushed my experimental implementation; see #417.

**Comment by pablogila** [comment]
This sounds like a good compromise.  

I am now testing the `qe` branch. The supercells have been created correctly, with the custom `H2` atoms. I will compare some calculations to test the new code, I will let you know how it goes.

**Comment by pablogila** [comment]
Ok, I spotted a potential source of trouble.  

The final file `phonopy.yaml` also had `H2`. In this case, this file cannot be read by other software (such as AbINS, etc). These programs usually read the atomic symbol and the mass, and use the mass to deduce different isotopes.  
For this reason, the ideal output would be (note the comments by `##`):  

```
  - symbol: H # 32        ##  'H' INSTEAD OF 'H2'
    coordinates: [  0.119266253800000,  0.341429274000000,  0.819720087700000 ]
    mass: 2.014100        ##  DEUTERIUM MASS
  - symbol: H  # 33       ##  'H'
    coordinates: [  0.550125137100000,  0.924319322700000,  0.324351909600000 ]
    mass: 1.007500        ##  PROTIUM MASS
```

However, with the current implementation from the `qe` branch, the output is actually

```
  - symbol: H2 # 32        ##  'H2' CAN NOT BE READ BY OTHER PROGRAMS OTHER THAN QE
    coordinates: [  0.119266253800000,  0.341429274000000,  0.819720087700000 ]
    mass: 2.014100         ##  DEUTERIUM MASS, CORRECT
  - symbol: H  # 33
    coordinates: [  0.550125137100000,  0.924319322700000,  0.324351909600000 ]
    mass: 1.007500
```
Ideally, the natural number after the atomic symbol should be removed when generating `phonopy.yaml`, leaving the custom mass to indicate to other programs that this is indeed an isotope. This would only be applied when generating `phonopy.yaml`; keeping `H2` when generating the supercells worked as intended.  

This is the only issue I could find. Replacing the `H2` with `H` via the scripts I linked earlier worked without further problems.

**Comment by atztogo** [comment]
Yaml is a structured format, therefore, the following is possible.
```
  - symbol: H # 32
    extended_symbol: H2
    coordinates: [  0.119266253800000,  0.341429274000000,  0.819720087700000 ]
    mass: 2.014100
  - symbol: H  # 33
    coordinates: [  0.550125137100000,  0.924319322700000,  0.324351909600000 ]
    mass: 1.007500
```

I don't want to add more definitions to phonopy.yaml to avoid increasing complexity, but it should be considered if it proves valuable enough.

**Comment by atztogo** [comment]
Pushed above PhonopyYaml format (`extended_symbol`).

**Comment by pablogila** [comment]
I have just tested the latest commit of the `qe` branch. It worked perfectly, the `phonopy.yaml` was easily read by external software, no bugs found. I don't think I have any further comments. Thanks for the implementation.
